--- a/Source/WebKit/qt/WebCoreSupport/QWebPageAdapter.cpp
+++ b/Source/WebKit/qt/WebCoreSupport/QWebPageAdapter.cpp
@@ -116,6 +116,10 @@
 #include "NetworkServicesClientQt.h"
 #endif
 
+#if !USE(SOUP)
+#include "CookieJarQt.h"
+#endif
+
 // from text/qfont.cpp
 QT_BEGIN_NAMESPACE
 extern Q_GUI_EXPORT int qt_defaultDpi();
@@ -377,6 +381,14 @@
 {
     if (!networkManager) {
         networkManager = new QNetworkAccessManager(handle());
+#if !USE(SOUP)
+        // set the shared cookieJar by default to the network manager
+        SharedCookieJarQt* cookieJar = SharedCookieJarQt::shared();
+        if(cookieJar) {
+          networkManager->setCookieJar(cookieJar);
+          cookieJar->setParent(0);
+      }
+#endif
     }
     return networkManager;
 }
